metabase/metabase
# Use column names instead of field refs for pivot viz settings

Epic https://github.com/metabase/metabase/issues/49405

There are 2 last viz settings that use field refs instead of column names:
```
{
  "pivot_table.column_split": {
    "rows": [["field", 1, null]],
    "columns": [["field", 2, null]],
    "values": [["aggregation", 0]],
  },
  "pivot_table.collapsed_rows": {
    "rows": [["field", 1, null]],
    "values": ["Gadget"],
  }
}
```

They cause the following issues:
- Applying a temporal unit parameter resets the viz settings since the ref changes (and fixing this is the reason for doing this work)
- When the QP switches between string and integer refs all pivot viz settings are lost
- Any change to field refs require us to write a custom migration or migrate data on read. Since we're reworking field refs now this can become a blocker.
- We cannot sync field ref changes when the query is updated (but we can with column names, e.g. `ID_2` -> `ID_3` changes)

This PR makes these settings work as others, where column names are used, which are synced with the query. Key parts of the solution:
- There is no migration. We support both old- and new-style viz settings both on the FE and the BE. It's not possible to write a custom migration for this since `result_metadata` is broken for pivot queries. It doesn't contain metadata for some of the returned columns, usually aggregations, but not only them.
- The FE uses new column name-based settings for new pivot questions. The FE doesn't change the settings for existing questions until they are explicitly modified, e.g. the columns are moved between rows/columns/values.
- Serialization supports both old and new-style settings. Column name settings are saved as is, while legacy settings are transformed according to the existing logic.
- Most tests were updated to use column names for viz settings to make sure we have enough coverage. There are still tests for the happy path and edge cases with field refs to make sure that legacy questions are supported.
- It's still an open question if we need to run a db migration or if that's already good enough. It's not possible to add a "custom migration" but we could use `quartz` in the future to rewrite cards in the background.

New settings look like this:
```
{
  "pivot_table.column_split": {
    "rows": ["CATEGORY"],
    "columns": ["VENDOR"],
    "values": ["sum"],
  },
  "pivot_table.collapsed_rows": {
    "rows": ["CATEGORY"],
    "values": ["Gadget"],
  }
}
```

How to verify:
- Checkout `master` and create pivot questions with several aggregations & breakouts. Make sure to test both default & custom column split (i.e. drag columns between sections in the viz settings)
- Checkout this branch and open your saved questions. There should be no changes. Your questions should not become ad-hoc.
- Modify pivot column split for some of the existing questions. Make sure that the settings have been migrated to column names and that it works as expected.
- Create new pivot questions. Make sure that column names are used by default.
